### ISSUE TITLE
Add compression support in WebSocket

### ISSUE THREAD

**Issue by zlatanov** [dotnet/runtime#31088]
Adds support for RFC 7692 "compression extensions for websocket".

## API Proposal

```c#
// new
public sealed class WebSocketCreationOptions
{
    // taken from existing WebSocket.CreateFromStream params.
    public bool IsServer { get; set; }
    public string? SubProtocol { get; set; }
    public TimeSpan KeepAliveInterval { get; set; }

    // new

    // turns the feature on.
    // consider: instead of multiple below properties on options class,
    // make a new DeflateCompressionOptions class and if null it is turned off.
    public bool UseDeflateCompression { get; set; } = false;

    // configures desired client window size (larger window = better compression, more memory usage)
    public int ClientMaxDeflateWindowBits { get; set; } = 15;

    // controls if the window is re-used between subsequent messages,
    // or if each message starts with a blank window.
    // memory vs compression tradeoff again.
    // might have a better name, "ClientPersistDeflateContext" etc.
    public bool ClientDeflateContextTakeover { get; set; } = true;

    public int ServerMaxDeflateWindowBits { get; set; } = 15;
    public bool ServerDeflateContextTakeover { get; set; } = true;
}

// existing
// this is set in ClientWebSocket.Options before calling Connect() against a URL.
public sealed class ClientWebSocketOptions
{
    // new, same as above.
    public bool UseDeflateCompression { get; set; } = false;
    public int ClientMaxDeflateWindowBits { get; set; } = 15;
    public bool ClientDeflateContextTakeover { get; set; } = true;
    public int ServerMaxDeflateWindowBits { get; set; } = 15;
    public bool ServerDeflateContextTakeover { get; set; } = true;
}

// existing
public abstract class WebSocket
{
    // existing
    public static WebSocket CreateFromStream(Stream stream, bool isServer, string? subProtocol, TimeSpan keepAliveInterval);

    // new
    public static WebSocket CreateFromStream(Stream stream, WebSocketCreationOptions options);
}
```

## Original Request

[AB#1118550](https://devdiv.visualstudio.com/10e66e43-9645-4201-b128-0fdc3769cc17/_workitems/edit/1118550)
See discussion here dotnet/runtime#20004.

At the moment the WebSocket doesn't support per-message deflate (see https://tools.ietf.org/html/rfc7692#section-7). Adding support for it in the BCL will mean that people (myself including) will no longer resort to implementing custom WebSockets in order to use it.

Proposed API
---
``` csharp
/// <summary>
/// Options to enable per-message deflate compression for <seealso cref="WebSocket" />.
/// </summary>
public sealed class WebSocketCompressionOptions
{
    /// <summary>
    /// This parameter indicates the base-2 logarithm of the LZ77 sliding window size of the client context.
    /// Must be a value between 8 and 15 or -1 indicating no preferences. The default is -1.
    /// </summary>
    public int ClientMaxWindowBits { get; set; } = -1;

    /// <summary>
    /// When true, the client informs the peer server of a hint that even if the server doesn't include the
    /// "client_no_context_takeover" extension parameter in the corresponding
    /// extension negotiation response to the offer, the client is not going to use context takeover. The default is false.
    /// </summary>
    public bool ClientNoContextTakeover { get; set; }

    /// <summary>
    /// This parameter indicates the base-2 logarithm of the LZ77 sliding window size of the server context.
    /// Must be a value between 8 and 15 or -1 indicating no preferences. The default is -1.
    /// </summary>
    public int ServerMaxWindowBits { get; set; } = -1;

    /// <summary>
    /// When true, the client prevents the peer server from using context takeover. If the peer server doesn't use context
    /// takeover, the client doesn't need to reserve memory to retain the LZ77 sliding window between messages. The default is false.
    /// </summary>
    public bool ServerNoContextTakeover { get; set; }
}

public sealed class ClientWebSocketOptions
{
    /// <summary>
    /// Instructs the <seealso cref="ClientWebSocket" /> to try and negotiate per-message compression.
    /// </summary>
    public WebSocketCompressionOptions Compression { get; set; }
}

public enum WebSocketOutputCompression
{
    /// <summary>
    /// Enables output compression if the underlying <seealso cref="WebSocket" /> supports it.
    /// </summary>
    Default,

    /// <summary>
    /// Suppresses output compression for the next message.
    /// </summary>
    SuppressOne,

    /// <summary>
    /// Suppresses output compression.
    /// </summary>
    Suppress
}

public abstract class WebSocket
{
    /// <summary>
    /// Instructs the socket to compress (or not to) the messages being sent when
    /// compression has been successfully negotiated.
    /// </summary>
    public WebSocketOutputCompression OutputCompression { get; set; } 

    public static WebSocket CreateFromStream(Stream stream, bool isServer, string subProtocol, TimeSpan keepAliveInterval, WebSocketCompressionOptions compression);
}
```

Rationale and Usage
---

The main drive behind the API is that we should not introduce any breaking changes. WebSockets already built will work as is. This is why I suggest we add new **CreateFromStream** method with `WebSocketCompressionOptions` parameter instead of adding it to the existing one.

There are a few options built in the WebSocket compression protocol that are considered advance use - controlling the size of the LZ77 sliding window, context takeover. We could easily hide them and choose reasonable defaults, but I think there are good use cases for them and as such we should expose them. See this blog post for good example of their usage: https://www.igvita.com/2013/11/27/configuring-and-optimizing-websocket-compression/.

The usage of the WebSocket would not change at all. By default a WebSocket created with compression options would compress all messages if the connection on the other end supports it. We introduce `OutputCompression` property to allow explicit opt out. The property has no effect if compression is not supported for the current connection.

Here is example of how we would disable compression for specific messages:
``` csharp
var socket = GetWebSocket();

// Disable compression for the next message only
socket.OutputCompression = WebSocketCompressionOptions.SuppressOne;
await socket.SendAsync(...);
```

Here is example for WebSocketClient:
``` csharp
var client = new WebSocketClient();

// Indicate that we want compression if server supports it
client.Options.Compression = new WebSocketCompressionOptions();

await client.ConnectAsync(...);

// Same as before. If compression is enabled it will be performed automatically.
client.SendAsync(...); 

// If we need to explicitly disable compression for a time or specific messages
client.OutputCompression = WebSocketOutputCompression.Suppress;

// Send one or more messages
// ...

// Restore default compression state
client.OutputCompression = WebSocketOutputCompression.Default; 
```

 Additional work will be required in AspNetCore repository and more specifically WebSocketMiddleware to light up the compression feature.

**Comment by davidfowl** [comment]
@zlatanov Are you still going to take this on?

cc @BrennanConroy

**Comment by zlatanov** [comment]
@davidfowl Yup.

**Comment by zlatanov** [comment]
@davidfowl @karelz Now that the main branch is 6.0, will this be considered? 

No feedback has been given since the API was proposed, almost 1 year ago (even just to say why it wasn't considered for 5.0 or at all). I'm sorry to say but it feels very off-putting.

**Comment by karelz** [comment]
@zlatanov yes, we plan to get back to this in 6.0. Sorry if it felt off-putting, but sadly, we have only limited bandwidth and even though this was important, it was less important than other features/APIs and we just didn't have anyone to do the research and provide valuable input.
This API was literally right below the cut line of 5.0 from the business perspective.

Given our current priorities (finishing off 5.0, YARP, QUIC), I think we should be able to get to this API in couple of months -- with plenty of time to finish it off in 6.0.
Let me know if you want to know more, or if I didn't fully answer your question / concerns. Thanks!

**Comment by jahmai** [comment]
@zlatanov I've been watching this for some time - look forward to you having the chance to get this out there 👍

**Comment by scalablecory** [comment]
Okay, all the API details proposed make sense now that I've skimmed the RFC.

I am not a web socket expert, so here are some thoughts -- @zlatanov @davidfowl does this make sense?:
- Is it valuable to keep the exact term `NoContextTakeover`? i.e. is it understood language that users would expect to carry across implementations? We may want to change this -- negative becomes positive `ClientContextTakeover`, or something a little more descriptive like `ClientPersistDeflateContext`.
- The `WebSocketOutputCompression` enum is a little weird -- it looks like we should put this in as a parameter to `SendAsync` instead.
- The memory savings are probably not important client-side, but I can see them being worthwhile server-side. I don't think it'll cost anything to enable for both, so we might as well, but if we want to be conservative we could probably leave the window/takeover settings off of `ClientWebSocketOptions`.

And general API review thoughts:
- To avoid adding even more constructors in future, we should consider creating a `WebSocketCreationOptions` class that is passed to `WebSocket.CreateFromStream` that we can add properties to in the future.
- Use nullables instead of -1 for window sizes.
- Fold the compression options into their parent types instead of introducing a new class.

So the final API would look something like:

```c#
// new
public sealed class WebSocketCreationOptions
{
    // taken from existing WebSocket.CreateFromStream params.
    public bool IsServer { get; set; }
    public string? SubProtocol { get; set; }
    public TimeSpan KeepAliveInterval { get; set; }

    // new
    public bool UseDeflateCompression { get; set; } = false;
    public int? ClientMaxDeflateWindowBits { get; set; }
    public bool ClientDeflateContextTakeover { get; set; } = true; // or ClientPersistDeflateContext etc.
    public int? ServerMaxDeflateWindowBits { get; set; }
    public bool ServerDeflateContextTakeover { get; set; } = true;
}

// existing
public sealed class ClientWebSocketOptions
{
    // new
    public bool UseDeflateCompression { get; set; } = false;
    public int? ClientMaxDeflateWindowBits { get; set; }
    public bool ClientDeflateContextTakeover { get; set; } = true; // or ClientPersistDeflateContext etc.
    public int? ServerMaxDeflateWindowBits { get; set; }
    public bool ServerDeflateContextTakeover { get; set; } = true;
}

// new
[Flags]
public enum WebSocketMessageFlags
{
    // taken from existing WebSocket.SendAsync params
    EndOfMessage = 1,

    // new
    Compress = 2
}

// existing
public abstract class WebSocket
{
    // existing
    public static WebSocket CreateFromStream (Stream stream, bool isServer, string? subProtocol, TimeSpan keepAliveInterval);
    public virtual ValueTask SendAsync(ReadOnlyMemory<byte> buffer, WebSocketMessageType messageType, bool endOfMessage, CancellationToken cancellationToken);

    // new
    public static WebSocket CreateFromStream(Stream stream, WebSocketCreationOptions options);
    public virtual ValueTask SendAsync(ReadOnlyMemory<byte> buffer, WebSocketMessageType messageType, WebSocketMessageFlags messageFlags, CancellationToken cancellationToken = default);
}
```

This appears straightforward to implement.

We can probably use the [existing deflate code](https://github.com/dotnet/runtime/tree/master/src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib) from System.IO.Compression. I think we'll want to add calls to `inflateReset` and `deflateReset` and maintain a pool of inflaters/deflaters for when the `NoContextTakeover` options are used, so we can realize the memory savings without going bonkers allocating new zlib contexts for every message. @carlossanlop @ericstj might have some guidance here.

**Comment by zlatanov** [comment]
@scalablecory thanks for the feedback. Here are my thoughts:

I completely agree that renaming some of the properties towards making more sense for consumers who are not (and probably should not) familiar with the RFC.

The implementation will be the same regardless server / client, but I agree that having this omitted from the client is a good thing, and basically matches what we have in the browsers right now where these options are handled by the browser itself.

Personally I am not fond of changing the API for sending messages. My reasoning is that consumers down the line don't need to be rebuild to support compression, one only need to change the setup of the websocket server / client. One other important fact about having a **WebSocketMessageFlags.Compress** flag in the API is that consumers must be aware when the socket really has compression enabled. For example a client might be created with compress options configured, but the server might not support compression or even it might decline it during the handshake. The opposite is true as well - the server might have compression enabled, but connected clients might not support it. 

I suggested WebSocketOutputCompression enum to be able to temporarily suppress compression for messages where it doesn't make sense, but I can agree that omitting such feature is not a big issue. Again it will align with the browser's implementations where once you have compression, all messages will be compressed.

**Comment by scalablecory** [comment]
> Personally I am not fond of changing the API for sending messages. My reasoning is that consumers down the line don't need to be rebuild to support compression, one only need to change the setup of the websocket server / client. One other important fact about having a `WebSocketMessageFlags.Compress` flag in the API is that consumers must be aware when the socket really has compression enabled. For example a client might be created with compress options configured, but the server might not support compression or even it might decline it during the handshake. The opposite is true as well - the server might have compression enabled, but connected clients might not support it.
> 
> I suggested `WebSocketOutputCompression` enum to be able to temporarily suppress compression for messages where it doesn't make sense, but I can agree that omitting such feature is not a big issue. Again it will align with the browsers' implementations where once you have compression, all messages will be compressed.

Do you have some specific examples of where you'd want to disable compression? I agree it'd be easy to just leave it off here, but if there's a compelling use case we should keep it.

If we do want to keep it, we can invert the feature -- instead of `WebSocketMessageFlags.Compress`, use `WebSocketMessageFlags.NoCompression`.

**Comment by zlatanov** [comment]
> Do you have some specific examples of where you'd want to disable compression? I agree it'd be easy to just leave it off here, but if there's a compelling use case we should keep it.
> 
> If we do want to keep it, we can invert the feature -- instead of `WebSocketMessageFlags.Compress`, use `WebSocketMessageFlags.NoCompression`. This way existing API users would just get the default of compress every message.

Of top of my head when I experimented with this (and it was quite long ago), very small messages or messages containing unique data such as timestamps or guids benefited very little or even had negative gain with compression enabled. If this remains area of interest, I can setup a few benchmarks to find out whether and when it's useful.

In the real world though, in all times I've had compression enabled, it was all or none, I didn't bother trying to measure which messages would be better off without compression because eventually message structures evolve and this kind of measurement is very brittle.

I do like your suggestion of inverting the feature and using `WebSocketMessageFlags.NoCompression` though.

**Comment by scalablecory** [comment]
> In the real world though, in all times I've had compression enabled, it was all or none, I didn't bother trying to measure which messages would be better off without compression because eventually message structures evolve and this kind of measurement is very brittle.

Okay. Easy thing to do here is nothing :). We can add that API later if someone finds a need.

**Comment by zlatanov** [comment]
@scalablecory I have one question about your suggested changes to the API proposal.

Isn't it better to have all compression options in the same object, rather than having them besides the existing ones. My reasoning is that it's not immediately obvious which options are related to the compression only, and having them configured without setting `UseDeflateCompression` is doing nothing.

```c#
public sealed class WebSocketCreationOptions
{
    public bool IsServer { get; set; }
    public string? SubProtocol { get; set; }
    public TimeSpan KeepAliveInterval { get; set; }

    // new
    public bool UseDeflateCompression { get; set; } = false;
    public int? ClientMaxDeflateWindowBits { get; set; } = null;
    public bool ClientDeflateContextTakeover { get; set; } = true; 
    public int? ServerMaxDeflateWindowBits { get; set; } = null;
    public bool ServerDeflateContextTakeover { get; set; } = true;
}
```
vs
``` c#
public sealed class WebSocketCreationOptions
{
    public bool IsServer { get; set; }
    public string? SubProtocol { get; set; }
    public TimeSpan KeepAliveInterval { get; set; }

    // new
    public WebSocketDefalateCompression? DeflateCompression { get; set; } 
}
```

In the second case we don't need `UseDeflateCompression` flag, because just having the object means we want compression. And the other benefit is that it's obvious what options towards compression are available. What are your thoughts?

**Comment by scalablecory** [comment]
We have some precedent for not grouping settings into their own objects, but I wouldn't mind it. I'll bring up both options during API review.

**Comment by davidfowl** [comment]
cc @BrennanConroy @halter73

**Comment by zlatanov** [comment]
@davidfowl What do you mean by that the client isn't aware? When a message is compressed, a bit flag is set in the message's headers. Also servers aren't supposed to send compressed messages if the handshake initiated by the client doesn't advocate that it wants it.

Other than that, if the server sends a compressed message, it's completely possible to read it as a normal message, and you will fail when trying to decode it as UTF-8 in case of a text message.

**Comment by scalablecory** [comment]
> What happens if the server sends compressed messages and the client isn't aware? Is it gracefully handled?

As @zlatanov said -- it is negotiated and uncompressed messages are part of spec, so client will handle it gracefully.

**Comment by BrennanConroy** [comment]
With regards to per-message compression:
We don't think SignalR would use a per-message compression flag as it would only apply to that transport and not the others. If we were to do some per message compression, it would be at a higher layer.

**Comment by terrajobst** [comment]
* Let's try to put the deflate settings in a single spot. This also solves the problem when we have to support a different compression algorithm with separate settings.

```C#
namespace System.Net.WebSockets
{
    public sealed class WebSocketDeflateOptions
    {
        public int ClientMaxWindowBits { get; set; } = 15;
        public bool ClientContextTakeover { get; set; } = true;
        public int ServerMaxWindowBits { get; set; } = 15;
        public bool ServerContextTakeover { get; set; } = true;
    }

    public sealed class WebSocketCreationOptions
    {
        public bool IsServer { get; set; }
        public string? SubProtocol { get; set; }
        public TimeSpan KeepAliveInterval { get; set; }
        public WebSocketDeflateOptions? DeflateOptions { get; set; } = null;
    }

    public partial class ClientWebSocketOptions
    {
        public WebSocketDeflateOptions? DeflateOptions { get; set; } = null;
    }

    public partial class WebSocket
    {
        public static WebSocket CreateFromStream(Stream stream, WebSocketCreationOptions options);
    }    
}
```

**Comment by karelz** [comment]
Thanks @scalablecory for shepherding the API proposal!

@zlatanov now that we have the API approved, are you interested in contributing the implementation? (I think you had an early prototype on GH)
If not, that is alright, we can take it on, just trying to avoid duplicate efforts ... Thanks!

**Comment by zlatanov** [comment]
@karelz Yes, I would love to do it.

**Comment by karelz** [comment]
That is great, thanks @zlatanov! I will find someone on our team to help with detailed code review.

**Comment by jogibear9988** [comment]
> * Let's try to put the deflate settings in a single spot. This also solves the problem when we have to support a different compression algorithm with separate settings.

If thats maybe a possible case, shouldn't we rename the property in  WebSocketCreationOptions to 

        public ICompressionOptions? CompressionOptions { get; set; } = null;

and DeflateOptions is a Instance of ICompressionOptions

**Comment by zlatanov** [comment]
@jogibear9988 This is discussed in the api review. See [here](https://www.youtube.com/watch?v=TPCjoYyp6e0&t=1h9m43s).

To summarize - it would complicate the api needlessly. If in the future other compression mechanisms become available, they will be exposed as separate options and it will be implementation detail which one to use if more than 1 of them is enabled.

**Comment by jogibear9988** [comment]
Thanks for the clarification.

**Comment by jogibear9988** [comment]
@zlatanov 
I've used this websocket before I switched to netcore: https://github.com/sta/websocket-sharp
This also has already compression. Maybe this is interesting, for example to create tests against or so.

**Comment by jogibear9988** [comment]
And one more question, will there also be a property to adjust the compression level & speed?

**Comment by zlatanov** [comment]
@jogibear9988 This is a good question. I always used the default options and never bothered with fastest or smallest as in my experience they never really differed much from the default for payload less than a few megabytes.

@scalablecory @terrajobst can say whether we should add `CompressionLevel` property. We can always add it later as it is a non breaking change.

**Comment by Benedicht** [comment]
Using the extension doesn't require to use it on all frames and compressing small payloads can be contraproductive, I would like to suggest a treshold parameter too.

**Comment by scalablecory** [comment]
> @scalablecory @terrajobst can say whether we should add `CompressionLevel` property. We can always add it later as it is a non breaking change.

We can add this later if needed; currently the need for this seems small. Please open a new issue for this so people can upvote if this is important to you.

> Using the extension doesn't require to use it on all frames and compressing small payloads can be contraproductive, I would like to suggest a treshold parameter too.

We considered an API for this -- not a threshold but simply a flag to selectively disable compression for each message -- but decided to hold on this as the benefit seemed trivially small. Please open a new issue for this so people can upvote if this is important to you.

**Comment by geoffkizer** [comment]
> We considered an API for this -- not a threshold but simply a flag to selectively disable compression for each message

A flag to enable/disable per message seems better than a threshold.

**Comment by karelz** [comment]
@zlatanov we have @CarnaViire on our team who will help with this. Please give her some time to ramp up as she is new to WebSockets.
She will start with the prereq #1493 to unblock you.
Do you have any preferred timeline in mind on your side? (just curious to make sure we are not off / too slow)
Thanks!

**Comment by zlatanov** [comment]
Thanks @karelz. No preferred timeline. Whenever the issue is completed/merged is fine.

**Comment by CarnaViire** [comment]
@zlatanov As #1493 is merged, could you please share with me your plans/estimates on this? Do you need any help?

**Comment by zlatanov** [comment]
@CarnaViire I have initial API changes ready and started working on the deflate implementation. I plan on completing it this weekend. After that I need to write several tests verifying the implementation which might take another week or so.

If you wish, before starting writing tests, I can share with you the branch where I am working for you to take a look and give me feedback?

**Comment by CarnaViire** [comment]
@zlatanov yes, it would be great if you could create a draft PR with your changes at that moment. You may additionally mark it as WIP if you wish.
It may take me some time to get through your changes (just setting expectations)

**Comment by CarnaViire** [comment]
Hey @zlatanov I understand that you might have been busy, but have you got any chance to proceed with the implementation? Is there anything I can help you with?

**Comment by zlatanov** [comment]
@CarnaViire I am working on the implementation, as it turned out little trickier than I thought because there are lots of assumptions inside the current implementation. Do you have somewhere where we could chat and figure out how we can collaborate?

**Comment by CarnaViire** [comment]
@zlatanov sure, just drop me an email (it should be visible on GH for now) and we'll figure it out from there.

**Comment by CarnaViire** [comment]
We've discussed that compression may cause security implications if turned on blindly. (You may see a part of that discussion on the PR here https://github.com/dotnet/runtime/pull/49304#issuecomment-792926408). We want to make sure that users understand that this is dangerous and encourage them to read the docs and weigh the risks. We also think it is important at this point to be able to turn off the compression for specific messages (if e.g. they contain a secret). That's why we want to add "Dangerous" prefix to the API and also add an API for per-message compression opt-out.

I've highlighted the changes we'd like to add to the proposal.

```diff
namespace System.Net.WebSockets
{
    public sealed class WebSocketDeflateOptions
    {
        public int ClientMaxWindowBits { get; set; } = 15;
        public bool ClientContextTakeover { get; set; } = true;
        public int ServerMaxWindowBits { get; set; } = 15;
        public bool ServerContextTakeover { get; set; } = true;
    }

    public sealed class WebSocketCreationOptions
    {
        public bool IsServer { get; set; }
        public string? SubProtocol { get; set; }
        public TimeSpan KeepAliveInterval { get; set; }
-       public WebSocketDeflateOptions? DeflateOptions { get; set; } = null;
+       public WebSocketDeflateOptions? DangerousDeflateOptions { get; set; } = null;
    }

    public partial class ClientWebSocketOptions
    {
-       public WebSocketDeflateOptions? DeflateOptions { get; set; } = null;
+       public WebSocketDeflateOptions? DangerousDeflateOptions { get; set; } = null;
    }

+   [Flags]
+   public enum WebSocketMessageFlags
+   {
+       // taken from existing WebSocket.SendAsync params
+       EndOfMessage = 1,
+
+       // new
+       DisableCompression = 2
+   }

    public partial class WebSocket
    {
        public static WebSocket CreateFromStream(Stream stream, WebSocketCreationOptions options);       
+       public virtual ValueTask SendAsync(ReadOnlyMemory<byte> buffer, WebSocketMessageType messageType, WebSocketMessageFlags messageFlags, CancellationToken cancellationToken = default);
    }    
}
```

cc @scalablecory

**Comment by danmoseley** [comment]
Would `EnableCompression` work as well as `DisableCompression` ? "Negative" sense flags can be confusing, eg "// set disable compression to false". Plus, it sounds like you want people to opt-in to it.

**Comment by CarnaViire** [comment]
@danmoseley our reason for naming it `DisableCompression`/`NoCompress`/whatever-opt-out was that you'd use that flag rarely. If you decide to use compression, you would want to use it for most messages, and only turn it off if you want to send a secret once in a while. If you send a lot of secrets, it's better not to use compression at all... And if you use the old `SendAsync` overload, where you don't have a flag, it will mean that compression is on. If we go with `EnableCompression`, the compression will be off by default, so we'll be forcing to use a new overload everywhere. It may not be such a bad thing though.

**Comment by zlatanov** [comment]
I like the added proposal changes. 

I don't think if you invert the flag to `EnableCompression` is a good choice. I think that it will pollute the existing code too much and it will give the impression that if I send a message with this flag on, the message will be compressed. This is not the case because it also depends whether compression has been enabled in the first place. This double enabling seems more confusing to me than to having to opt out.

**Comment by danmoseley** [comment]
I defer to whatever API review says ☺

**Comment by CarnaViire** [comment]
We've discussed the things left for consideration.

_1) Whether the "Dangerous" prefix might be overused so it will lose its purpose_
We agreed on keeping the dangerous prefix. We think the danger is completely non-obvious so it's good to call it out. We don't think enough people will use it to dilute the prefix's meaning.

_2) Defaulting `WebSocketMessageFlags` parameter in `SendAsync`_
We don't think `None` will be a good default. It seems likely that you're either calling `EndOfMessage` every time, or are calling it some of the time... but definitely not never. `EndOfMessage` is also not a good default, because then it will be counter-intuitive to call `SendAsync` explicitly with `None` to state the message will have a continuation. That's why we believe it makes sense to ask the caller to always specify the flag, no defaulting.

If anyone sees any issues with our decisions, please let us know.

**Comment by danmoseley** [comment]
Is it possible that there might in future be a solid mitigation to this dangerousness? Eg., a form of compression that pads to a fixed length. In that case, might we have a WebSocketDeflateOptions member that wasn't dangerous, and yet was accessed with DangerousDeflateOptions ?

**Comment by geoffkizer** [comment]
> Is it possible that there might in future be a solid mitigation to this dangerousness? Eg., a form of compression that pads to a fixed length. In that case, might we have a WebSocketDeflateOptions member that wasn't dangerous, and yet was accessed with DangerousDeflateOptions ?

I doubt it. For compression to be useful, it has to compress to something smaller than the original. If it does, then it seems like it's still subject to a CRIME attack.

**Comment by geoffkizer** [comment]
To be clear, the general mitigation is to ensure you don't use the same compression context to compress data from different sources, at least one of which is untrusted. Unfortunately it's not easy to figure out how to enforce that generally.
